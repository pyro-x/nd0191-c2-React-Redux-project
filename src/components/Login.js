import { useState } from 'react';
import { connect } from 'react-redux';
import { setAuthedUser } from '../actions/authedUser';
import { useNavigate } from 'react-router-dom';

const Login = (props) => {

    const { users , dispatch} = props;

    let [username, setUsername] = useState('');
    let [password, setPassword] = useState('');
    let [errorMsg, setErrorMsg] = useState('');

    let [avatar, setAvatar] = useState('unknown-avatar.svg');
    const navigate =useNavigate();

    const handleSubmit = (e) => {
        e.preventDefault();
        if (username === '' || password === '') {
            setErrorMsg('Please enter a username and password');
            return;
        }
        const selectedUser = users.filter (user => user.id === username)[0];

        if (selectedUser.password === password) {

            setErrorMsg('');
            dispatch(setAuthedUser(username));
            navigate(-1);

        } else {
            setErrorMsg('Invalid password');
        }
    }

    const handleUserNameChange = (e) => {
        setUsername(e.target.value);
        setAvatar(users.filter (user => user.id === e.target.value)[0].avatarURL);
    }

    const handlePaswordChange = (e) => {
        setPassword(e.target.value);
    }

    return (
        <div className='border shadow-md'>
            <form className='flex flex-col space-y-4 m-4' onSubmit={handleSubmit}>
                <img src={avatar} alt={username ? username : 'Mistery Avatar'}  data-testid="user-img" className='w-32 h-32 self-center' />
                <div className='flex justify-between items-baseline'>
                    <div className="pr-2">User</div> 
                    <select className="w-64" name="users" id="users" data-testid="user-select"  value={username} onChange={handleUserNameChange}>
                        <option  value="" defaultValue="a">Please Select a user</option>
                        {
                            users.map((user) => (
                                <option key={user.id} value={user.id}>{user.name}</option>
                            ))
                        }
                    </select>
                </div>
                <div className='flex justify-between items-baseline'>
                    <div className='pr-2'>Password</div> 
                    <input className="w-64" type="password" placeholder="password"  data-testid="password-input" onChange={handlePaswordChange} value={password}/>
                </div>
            { errorMsg !=='' && <div className="text-red-500 text-sm text-right" data-testid="error-msg">{errorMsg}</div>}
                <button   type="submit" className="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded-md mt-8 disabled:opacity-50" disabled={username === ""} data-testid="submit-button">Login</button>

            </form>
        </div>
    )
}
const mapStateToProps = ({ users }) => {
    return {
        users: Object.values(users).map ((user) => {
            return {
                id: user.id,
                name: user.name,
                password: user.password,
                avatarURL: user.avatarURL
            }
        }),
    }
}

export default connect(mapStateToProps)(Login);