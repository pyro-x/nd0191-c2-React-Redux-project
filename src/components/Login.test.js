import Login from './Login';

import { fireEvent, render, screen } from '@testing-library/react';

import * as React from 'react';

import reducer from "../reducers";
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import middleware from "../middleware";
import { createStore } from 'redux';
import { MemoryRouter } from 'react-router';




describe ('Login', () => {
    let store;
    const initialState = {
        authedUser: null,
        users: {
            sarahedo: {
                id: "sarahedo",
                name: "Sarah Edo",
                avatarURL: "https://api.dicebear.com/5.x/avataaars/svg?seed=sarahedo&style=circle",
                answers: {
                    "8xf0y6ziyjabvozdd253nd": "optionOne",
                    "6ni6ok3ym7mf1p33lnez": "optionOne",
                    "am8ehyc8byjqgar0jgpub9": "optionTwo",
                    "loxhs1bqm25b708cmbf3g": "optionTwo"
                },
                questions: [],
                password: "password123"
            }
        }
    }

    beforeEach(() => {
        store = createStore(reducer, initialState, middleware);
    });

    it ('renders correctly', () => {
        const  view  = render(
            <MemoryRouter>
                <Provider store={store}>
                    <Login/>
                </Provider>
            </MemoryRouter>
            );
        expect(view).toMatchSnapshot()
    });

    it ('user select password and submit button are present', () => {
        const  view  = render(
            <MemoryRouter>
                <Provider store={store}>
                    <Login/>
                </Provider>
            </MemoryRouter>
            );
        expect(screen.getByTestId('user-select')).toBeInTheDocument();
        expect(screen.getByTestId('password-input')).toBeInTheDocument();
        expect(screen.getByTestId('submit-button')).toBeInTheDocument();
    });


    it ('img logo changes when user selected', () => {
        const  view  = render(
            <MemoryRouter>
                <Provider store={store}>
                    <Login/>
                </Provider>
            </MemoryRouter>
            );
        
        var input = screen.getByTestId('user-select');
        fireEvent.change (input, {target: {value: "sarahedo"}});
        expect (screen.getByTestId('user-img')).toHaveAttribute('src', 'https://api.dicebear.com/5.x/avataaars/svg?seed=sarahedo&style=circle');
        expect(screen.getByTestId('user-img')).toHaveAttribute('alt', 'sarahedo');

    });

    it ('shows password error when wrong password entered', () => {
        const  view  = render(
            <MemoryRouter>
                <Provider store={store}>
                    <Login/>
                </Provider>
            </MemoryRouter>
            );

        var input = screen.getByTestId('user-select');
        var password = screen.getByTestId('password-input');
        fireEvent.change (input, {target: {value: "sarahedo"}});
        fireEvent.change (password, {target: {value: "wrongpassword"}}); // wrong password

        fireEvent.click(screen.getByTestId('submit-button'));
        expect(screen.getByTestId('error-msg')).toBeInTheDocument();

    });
}); 