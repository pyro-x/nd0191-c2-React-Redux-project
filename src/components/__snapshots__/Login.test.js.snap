// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="border shadow-md"
      >
        <form
          class="flex flex-col space-y-4 m-4"
        >
          <img
            alt="Mistery Avatar"
            class="w-32 h-32 self-center"
            data-testid="user-img"
            src="unknown-avatar.svg"
          />
          <div
            class="flex justify-between items-baseline"
          >
            <div
              class="pr-2"
            >
              User
            </div>
            <select
              class="w-64"
              data-testid="user-select"
              id="users"
              name="users"
            >
              <option
                value=""
              >
                Please Select a user
              </option>
              <option
                value="sarahedo"
              >
                Sarah Edo
              </option>
            </select>
          </div>
          <div
            class="flex justify-between items-baseline"
          >
            <div
              class="pr-2"
            >
              Password
            </div>
            <input
              class="w-64"
              data-testid="password-input"
              placeholder="password"
              type="password"
              value=""
            />
          </div>
          <button
            class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded mt-8 disabled:opacity-50"
            data-testid="submit-button"
            disabled=""
            type="submit"
          >
            Login
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="border shadow-md"
    >
      <form
        class="flex flex-col space-y-4 m-4"
      >
        <img
          alt="Mistery Avatar"
          class="w-32 h-32 self-center"
          data-testid="user-img"
          src="unknown-avatar.svg"
        />
        <div
          class="flex justify-between items-baseline"
        >
          <div
            class="pr-2"
          >
            User
          </div>
          <select
            class="w-64"
            data-testid="user-select"
            id="users"
            name="users"
          >
            <option
              value=""
            >
              Please Select a user
            </option>
            <option
              value="sarahedo"
            >
              Sarah Edo
            </option>
          </select>
        </div>
        <div
          class="flex justify-between items-baseline"
        >
          <div
            class="pr-2"
          >
            Password
          </div>
          <input
            class="w-64"
            data-testid="password-input"
            placeholder="password"
            type="password"
            value=""
          />
        </div>
        <button
          class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded mt-8 disabled:opacity-50"
          data-testid="submit-button"
          disabled=""
          type="submit"
        >
          Login
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
